### PR TITLE
libbpf: 0.8.1 -> 1.0.0

### PR DESCRIPTION
**Pull request by saschagrunert**
###### Description of changes

Update libbpf to the latest release.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by saschagrunert** [comment]
cc @thoughtpolice @vcunat @martinetd

**Comment by vcunat** [comment]
Please wait with this.

> All deprecated APIs and features removed!

This means IIRC that what they moved to libxdp is gone now, but we don't package libxdp yet.  Example breakage: `knot-dns` build.

**Comment by vcunat** [comment]
Cross-ref: https://github.com/xdp-project/xdp-tools/issues/220

**Comment by martinetd** [comment]
Thanks @vcunat! It looks like you're working on libxdp (they were fast fixing that particular error!) so leaving follow up to you, but please ask if there's anything actionable for others :)

**Comment by vcunat** [comment]
Multiple packages have issues building, apparently:
- [ ] `knot-dns`: that's for myself and upstream colleagues, WIP
- [ ] `pahole`
- [ ] `bcc`
- [ ] `bpftrace`? (hidden by `bcc`)
- [ ] `suricata`

**Comment by martinetd** [comment]
- bcc -> ah, that'd be why upstream released 0.25.0 probably... unfortunately that breaks bpftrace in turn; this hasn't been fixed upstream and I'll get around to it eventually but I didn't go as far as to check libbpf 1.0...
- pahole -> I'll have a look

**Comment by martinetd** [comment]
pahole 1.24 got released last week and builds fine with libbpf 1.0.0 -- on the other hand it does not build with libbpf 0.8.0 (btf.h too old) so we need to upgrade both together.

@saschagrunert can you pick https://github.com/NixOS/nixpkgs/compare/master...martinetd:nixpkgs:libbpf_1.0 up ?

**Comment by saschagrunert** [comment]
Rebased on top of the latest `master` branch.

**Comment by vcunat** [comment]
@martinetd: the point was to add the pahole update which is *not* in nixpkgs master.

**Comment by martinetd** [comment]
(Sorry I meant to link to https://github.com/martinetd/nixpkgs/commits/libbpf_1.0 -- doesn't change much anyway; as pointed out that's not in master yet because it wouldn't work here on its own)


I can also confirm bcc 0.25.0 works with libbpf 1.0.0 -- this one also compiles with libbpf 0.8.1 (positively surprised here!) so I can open a separate PR for it as soon as I got bpftrace sorted out to work with it; although we can also group it together with this I have no preference.

(I'd also like to upgrade the llvm toolchain used for bcc/bpftrace while we're doing this -- llvm 11 is getting a bit old and while it appears to work I'd be more comfortable with llvm13... and -that- definitely belongs in a different PR... It's just a matter of specifying `llvmPackages = llvmPackages_13` in all-packages. But as long as the newer version works I see no reason to push for this here; will do that after bpftrace fix)

**Comment by vcunat** [comment]
I prefer to decouple PRs, unless there's some particular reason.  (generally and here I can't see any such reason)

**Comment by martinetd** [comment]
makes sense to me.
for bpftrace: https://github.com/iovisor/bpftrace/issues/2340
tl;dr -- the master branch there does build with bcc 0.25.0 as they removed the dependency on the problematic function, but it's a huge step from bpftrace 0.15.0 so I'm a bit reluctant to go with it unless they cut a new version tbh.
I also have a patch that fixes it for 0.15 before checking master branch was completely different, so we can go with that if that looks better...

Either approach should work fine with older libbpf as well, so no problem with doing that in another PR -- I'll open one once I get a reply from upstream about a new release.

**Comment by veehaitch** [comment]
It seems, systemd has no support for `libbpf` v1.0.0 yet:

https://github.com/systemd/systemd/blob/1a0e065e9f154f46fd68cd45f46310bc7df7a51c/src/shared/bpf-dlopen.c#L49

**Comment by martinetd** [comment]
> It seems, systemd has no support for libbpf v1.0.0 yet:

Good point; I had a quick look at the symbol list and it's not just bumping the soname: `bpf_create_map`, `bpf_map__resize`, and `bpf_probe_prog_type` got removed.

Can you/do you want to open an issue in their repo?


In the meantime I'm starting to think we might want to just add libbpf 1.0.0 as a new package and let people pick one, that should allow a smoother transition....

**Comment by vcunat** [comment]
> In the meantime I'm starting to think we might want to just add libbpf 1.0.0 as a new package and let people pick one

Yes.  Or at least when someone appears who can benefit from upgrading to >= 1.0.0.

**Comment by martinetd** [comment]
bpftrace 0.16.0 got released just now so opened https://github.com/NixOS/nixpkgs/pull/189078 for it. It works with older libbpf as expected, so can be done separately and we'll just have to rebase.

That leaves knot-dns, systemd (just replied in that issue, thanks @veehaitch ) and.. suricata ?
@magenbluten (suricata maintainer) as he wasn't part of this issue until now.
Unless we want to just leave both as previously suggested and move forward; I don't need anything in libbpf 1.0 at this point.

**Comment by martinetd** [comment]
closing as https://github.com/NixOS/nixpkgs/pull/193591 got merged

**Comment by martinetd** [comment]
Update for the dependencies above: https://github.com/NixOS/nixpkgs/pull/187978#issuecomment-1227122103
- suricatta got merged in another PR a couple of weeks ago, but only in their master branch so that'll require a backport to their 6.x branch if we want an update; I don't think there's any hurry though so won't bother asking
- systemd merged my PR and it's in v252-rc1; they plan on backporting it to stable after it's had some testing on master (probably after release?) so we'll get it sooner or later as well now

I don't think there's any hurry for these; but I'd like to get pahole in ( https://github.com/NixOS/nixpkgs/pull/194551 ) as it's had some nice performance improvement through parallelism of the BTF encoder if one of you could take a look :)
